zokrates_core/src/static_analysis/zir_propagation.rs

```diff
@@ -0,0 +1,1694 @@
+};
+use std::collections::HashMap;
+use std::fmt;
+use zokrates_field::Field;
+
+type Constants<'ast, T> = HashMap<Identifier<'ast>, ZirExpression<'ast, T>>;
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    OutOfBounds(u128, u128),
+    DivisionByZero,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Error::OutOfBounds(index, size) => write!(
+                f,
+                "Out of bounds index ({} >= {}) found in zir during static analysis",
+                index, size
+            ),
+            Error::DivisionByZero => {
+                write!(f, "Division by zero detected in zir during static analysis",)
+            }
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct ZirPropagator<'ast, T> {
+    constants: Constants<'ast, T>,
+}
+
+impl<'ast, T: Field> ZirPropagator<'ast, T> {
+    pub fn propagate(p: ZirProgram<T>) -> Result<ZirProgram<T>, Error> {
+        ZirPropagator::default().fold_program(p)
+    }
+}
+
+impl<'ast, T: Field> ResultFolder<'ast, T> for ZirPropagator<'ast, T> {
+    type Error = Error;
+
+    fn fold_statement(
+        &mut self,
+        s: ZirStatement<'ast, T>,
+    ) -> Result<Vec<ZirStatement<'ast, T>>, Self::Error> {
+        match s {
+            ZirStatement::Assertion(e) => match self.fold_boolean_expression(e)? {
+                BooleanExpression::Value(true) => Ok(vec![]),
+                e => Ok(vec![ZirStatement::Assertion(e)]),
+            },
+            ZirStatement::Definition(a, e) => {
+                let e = self.fold_expression(e)?;
+                match e {
+                    ZirExpression::FieldElement(FieldElementExpression::Number(..))
+                    | ZirExpression::Boolean(BooleanExpression::Value(..))
+                    | ZirExpression::Uint(UExpression {
+                        inner: UExpressionInner::Value(..),
+                        ..
+                    }) => {
+                        self.constants.insert(a.id, e);
+                        Ok(vec![])
+                    }
+                    _ => {
+                        self.constants.remove(&a.id);
+                        Ok(vec![ZirStatement::Definition(a, e)])
+                    }
+                }
+            }
+            ZirStatement::MultipleDefinition(assignees, list) => {
+                for a in &assignees {
+                    self.constants.remove(&a.id);
+                }
+                Ok(vec![ZirStatement::MultipleDefinition(
+                    assignees,
+                    self.fold_expression_list(list)?,
+                )])
+            }
+            _ => fold_statement(self, s),
+        }
+    }
+
+    fn fold_field_expression(
+        &mut self,
+        e: FieldElementExpression<'ast, T>,
+    ) -> Result<FieldElementExpression<'ast, T>, Self::Error> {
+        match e {
+            FieldElementExpression::Number(n) => Ok(FieldElementExpression::Number(n)),
```

Comment:
this looks like the folder's default, could be removed and covered by a single arm calling the default, see other optimization steps

Reply:
ok it's short enough like this, closing